File: Word-Break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
            
        output = False
        
        cache = {}
        
        def rec(index):
            
            nonlocal output
            
            tmp = False
            
            if index == len(s):
                
                output = True
                
                return True
            
            if index in cache:
                
                return cache[index]
        
            for each in wordDict:
                
                if each == s[index:index+len(each)]:
                    
                    tmp = rec(index+len(each))
                    
            cache[index] = tmp
                    
            return tmp
        
        rec(0)
        
        return output

File: Word-Break.java
class Solution {
    
    String s;
    List<String> wordDict;
    
    HashMap<Integer,Boolean> cache = new HashMap<>();    
    
    public boolean rec(int index) {
        
        if (index == s.length()) {
                                    
            return true;
        }
        
        if (index > s.length()) {
            
            return false;
        }
        
        if (cache.containsKey(index)) {
            
            return cache.get(index);
        }
        
        Boolean ans = false;
        
        for (String each:wordDict) {
            
            if ((index+each.length()) <= s.length()) {
            
                if (s.substring(index,index+each.length()).equals(each)){

                    ans = ans || rec(index+each.length());

                }
                
            }
            
        }
        
        cache.put(index,ans);
        
        return ans;
        
    }
    
    public boolean wordBreak(String s, List<String> wordDict) {
        
        this.s = s;
        this.wordDict = wordDict;
        
        return rec(0);
        
    }
}
